drivers: sensor: ina23x: Use micro-ohms for rshunt.

bperseghetti opened the pull request:
Changes rshunt-milliohms to rshunt-microohms allowing for current sensing of greater than 16.4A (1mOhm resistor). This is commonly set to 100 uOhm for VMU/FMU boards/applications.

Fixes #59862

bperseghetti committed:
drivers: sensor: ina23x: Use micro-ohms for rshunt

Changes rshunt-milliohms to rshunt-micro-ohms allowing for current
sensing of greater than 16.4A (1mOhm resistor). This is commonly
set to 100 uOhm for VMU/FMU boards/applications.

Co-authored-by: James Goppert <james.goppert@gmail.com>
Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

bbilas commented:
@bperseghetti You are changing dts bindings so that change should be noted in Release Notes.

gmarull commented on dts/bindings/sensor/ti,ina23x-common.yaml:
```diff
@@ -34,10 +34,10 @@ properties:
       so that low maximum currents can be measured with enough resolution.
 
-  rshunt-milliohms:
+  rshunt-microohms:
```
```suggestion
  rshunt-micro-ohms:
```

per dtschema convention

bperseghetti replied:
Thanks @gmarull, Changes made.

bperseghetti commented:
> @bperseghetti You are changing dts bindings so that change should be noted in Release Notes.

Added it to the release notes for 3.5. Thanks for pointing that out, I followed the same format for when they were changed from rshunt to rshunt-milliohm in release 3.3 notes.

gmarull commented on doc/releases/release-notes-3.5.rst:
```diff
@@ -233,6 +233,23 @@ USB
 Devicetree
 **********
 
+Bindings
+========
+
+Modified bindings:
+
+  - Texas Instruments (ti):
+
+    - :dtcompatible:`ti,ina230` (on i2c bus):
+
+        - new property: ``rshunt-micro-ohms``
+        - removed property: ``rshunt-milliohms``
+
+    - :dtcompatible:`ti,ina237` (on i2c bus):
+
+        - new property: ``rshunt-micro-ohms``
+        - removed property: ``rshunt-milliohms``
```
this is autogenerated before the release, please drop

bperseghetti replied:
Removed.

bperseghetti commented:
@bbilas could you also approve the PR if you don't see any other needed changes?

bperseghetti commented:
Thanks @bbilas @gmarull ! Is this now able to be merged in?

xyzzy42 commented on drivers/sensor/ina23x/ina230.c:
```diff
@@ -192,7 +192,7 @@ static int ina230_calibrate(const struct device *dev)
 	/* See datasheet "Programming" section */
 	val = (INA230_CAL_SCALING * 10000U) /
-	      (config->current_lsb * config->rshunt);
+	      ((config->current_lsb * config->rshunt) / 1000U);
```
All the values here are build time constants.  Instead of rounding and making the value less accurate, you can just compute at compile time with 64-bit math.

bperseghetti replied:
Thanks for the input, I tried this with a few different methods, but note that it needs to be calcualted per each instance from the device tree, therefore making it required to be in the config structure. (please see ina230x.patch below)

I originally went down this road to try and get a  uint16_t .cal option created for each dt instance.
IE:
```c
.cal = INA230_CAL_SCALING * 1000000000U /	    \
			(DT_INST_PROP(inst, current_lsb_microamps) *\
			DT_INST_PROP(inst, rshunt_micro_ohms)),     \
```

I then realized that in real tests this method also had issues with precision and after changing it to:

ina230x.patch:
```
diff --git a/drivers/sensor/ina23x/ina230.c b/drivers/sensor/ina23x/ina230.c
index 6e7c23ba15..f3cc9486ae 100644
--- a/drivers/sensor/ina23x/ina230.c
+++ b/drivers/sensor/ina23x/ina230.c
@@ -187,14 +187,10 @@ static int ina230_attr_get(const struct device *dev, enum sensor_channel chan,
 static int ina230_calibrate(const struct device *dev)
 {
 	const struct ina230_config *config = dev->config;
-	uint16_t val;
 	int ret;
 
 	/* See datasheet "Programming" section */
-	val = (INA230_CAL_SCALING * 10000U) /
-	      ((config->current_lsb * config->rshunt) / 1000U);
-
-	ret = ina23x_reg_write(&config->bus, INA230_REG_CALIB, val);
+	ret = ina23x_reg_write(&config->bus, INA230_REG_CALIB, config->cal);
 	if (ret < 0) {
 		return ret;
 	}
@@ -276,7 +272,9 @@ static const struct sensor_driver_api ina230_driver_api = {
 		.bus = I2C_DT_SPEC_INST_GET(inst),		    \
 		.config = DT_INST_PROP(inst, config),		    \
 		.current_lsb = DT_INST_PROP(inst, current_lsb_microamps),\
-		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),	    \
+		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),    \
+		.cal = 5120000U / ((DT_INST_PROP(inst, current_lsb_microamps) *\
+			DT_INST_PROP(inst, rshunt_micro_ohms) / 1000U)),\
 		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, alert_gpios),\
 			    (INA230_CFG_IRQ(inst)), ())		    \
 	};							    \
diff --git a/drivers/sensor/ina23x/ina230.h b/drivers/sensor/ina23x/ina230.h
index 68a4bbef23..63955ca6ff 100644
--- a/drivers/sensor/ina23x/ina230.h
+++ b/drivers/sensor/ina23x/ina230.h
@@ -51,6 +51,7 @@ struct ina230_config {
 	uint16_t config;
 	uint32_t current_lsb;
 	uint32_t rshunt;
+	uint16_t cal;
 #ifdef CONFIG_INA230_TRIGGER
 	bool trig_enabled;
 	uint16_t mask;
diff --git a/drivers/sensor/ina23x/ina237.c b/drivers/sensor/ina23x/ina237.c
index 468f4af9f9..fe7485bb1d 100644
--- a/drivers/sensor/ina23x/ina237.c
+++ b/drivers/sensor/ina23x/ina237.c
@@ -244,10 +244,7 @@ static int ina237_calibrate(const struct device *dev)
 	int ret;
 
 	/* see datasheet "Current and Power calculations" section */
-	val = (INA237_CAL_SCALING * ((config->current_lsb * config->rshunt) /
-	       1000U)) / 10000000U;
-
-	ret = ina23x_reg_write(&config->bus, INA237_REG_CALIB, val);
+	ret = ina23x_reg_write(&config->bus, INA237_REG_CALIB, config->cal);
 	if (ret < 0) {
 		return ret;
 	}
@@ -379,7 +376,9 @@ static const struct sensor_driver_api ina237_driver_api = {
 		.config = DT_INST_PROP(inst, config),				\
 		.adc_config = DT_INST_PROP(inst, adc_config),			\
 		.current_lsb = DT_INST_PROP(inst, current_lsb_microamps),	\
-		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),			\
+		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),		\
+		.cal = 8192U * (DT_INST_PROP(inst, current_lsb_microamps) *	\
+			DT_INST_PROP(inst, rshunt_micro_ohms) / 1000U) / 10000000U,\
 		.alert_config = DT_INST_PROP_OR(inst, alert_config, 0x01),	\
 		.alert_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, alert_gpios, {0}),	\
 	};							    \
diff --git a/drivers/sensor/ina23x/ina237.h b/drivers/sensor/ina23x/ina237.h
index b028f5ffb9..e923bbaf05 100644
--- a/drivers/sensor/ina23x/ina237.h
+++ b/drivers/sensor/ina23x/ina237.h
@@ -50,6 +50,7 @@ struct ina237_config {
 	uint16_t adc_config;
 	uint32_t current_lsb;
 	uint32_t rshunt;
+	uint16_t cal;
 	const struct gpio_dt_spec alert_gpio;
 	uint16_t alert_config;
 };
```

I still noticed that it didn't pass clang formatting. Since this doesn't seem to be providing any extra value and I'm performing the correct order of operations any, I think I will leave the current PR the way it is.

xyzzy42 replied:
The key is that you can use 64-bit math for free here, since it will all be resolved at compile time.

```c
.cal = (uint16_t)((INA230_CAL_SCALING * 10000000ULL) / \
        ((uint64_t)DT_INST_PROP(inst, current_lsb_microamps) * \
        DT_INST_PROP(inst, rshunt_micro_ohms))), \
```

Now there are no longer issues with precision nor overflow.

xyzzy42 commented on drivers/sensor/ina23x/ina237.c:
```diff
@@ -244,8 +244,8 @@ static int ina237_calibrate(const struct device *dev)
-	val = (INA237_CAL_SCALING * config->current_lsb * config->rshunt) /
-	       10000000U;
+	val = (INA237_CAL_SCALING * ((config->current_lsb * config->rshunt) /
+	       1000U)) / 10000000U;
```
This is a less accurate than it needs to be.   And again, it can be computed at build time with 64-bit math for free.

bperseghetti replied:
Thanks for the input, I tried this with a few different methods, but note that it needs to be calcualted per each instance from the device tree, therefore making it required to be in the config structure. (please see ina230x.patch below)

I originally went down this road to try and get a  uint16_t .cal option created for each dt instance.
IE:
```c
.cal = INA230_CAL_SCALING * 1000000000U /	    \
			(DT_INST_PROP(inst, current_lsb_microamps) *\
			DT_INST_PROP(inst, rshunt_micro_ohms)),     \
```

I then realized that in real tests this method also had issues with precision and after changing it to:

ina230x.patch:
```
diff --git a/drivers/sensor/ina23x/ina230.c b/drivers/sensor/ina23x/ina230.c
index 6e7c23ba15..f3cc9486ae 100644
--- a/drivers/sensor/ina23x/ina230.c
+++ b/drivers/sensor/ina23x/ina230.c
@@ -187,14 +187,10 @@ static int ina230_attr_get(const struct device *dev, enum sensor_channel chan,
 static int ina230_calibrate(const struct device *dev)
 {
 	const struct ina230_config *config = dev->config;
-	uint16_t val;
 	int ret;
 
 	/* See datasheet "Programming" section */
-	val = (INA230_CAL_SCALING * 10000U) /
-	      ((config->current_lsb * config->rshunt) / 1000U);
-
-	ret = ina23x_reg_write(&config->bus, INA230_REG_CALIB, val);
+	ret = ina23x_reg_write(&config->bus, INA230_REG_CALIB, config->cal);
 	if (ret < 0) {
 		return ret;
 	}
@@ -276,7 +272,9 @@ static const struct sensor_driver_api ina230_driver_api = {
 		.bus = I2C_DT_SPEC_INST_GET(inst),		    \
 		.config = DT_INST_PROP(inst, config),		    \
 		.current_lsb = DT_INST_PROP(inst, current_lsb_microamps),\
-		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),	    \
+		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),    \
+		.cal = 5120000U / ((DT_INST_PROP(inst, current_lsb_microamps) *\
+			DT_INST_PROP(inst, rshunt_micro_ohms) / 1000U)),\
 		COND_CODE_1(DT_INST_NODE_HAS_PROP(inst, alert_gpios),\
 			    (INA230_CFG_IRQ(inst)), ())		    \
 	};							    \
diff --git a/drivers/sensor/ina23x/ina230.h b/drivers/sensor/ina23x/ina230.h
index 68a4bbef23..63955ca6ff 100644
--- a/drivers/sensor/ina23x/ina230.h
+++ b/drivers/sensor/ina23x/ina230.h
@@ -51,6 +51,7 @@ struct ina230_config {
 	uint16_t config;
 	uint32_t current_lsb;
 	uint32_t rshunt;
+	uint16_t cal;
 #ifdef CONFIG_INA230_TRIGGER
 	bool trig_enabled;
 	uint16_t mask;
diff --git a/drivers/sensor/ina23x/ina237.c b/drivers/sensor/ina23x/ina237.c
index 468f4af9f9..fe7485bb1d 100644
--- a/drivers/sensor/ina23x/ina237.c
+++ b/drivers/sensor/ina23x/ina237.c
@@ -244,10 +244,7 @@ static int ina237_calibrate(const struct device *dev)
 	int ret;
 
 	/* see datasheet "Current and Power calculations" section */
-	val = (INA237_CAL_SCALING * ((config->current_lsb * config->rshunt) /
-	       1000U)) / 10000000U;
-
-	ret = ina23x_reg_write(&config->bus, INA237_REG_CALIB, val);
+	ret = ina23x_reg_write(&config->bus, INA237_REG_CALIB, config->cal);
 	if (ret < 0) {
 		return ret;
 	}
@@ -379,7 +376,9 @@ static const struct sensor_driver_api ina237_driver_api = {
 		.config = DT_INST_PROP(inst, config),				\
 		.adc_config = DT_INST_PROP(inst, adc_config),			\
 		.current_lsb = DT_INST_PROP(inst, current_lsb_microamps),	\
-		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),			\
+		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),		\
+		.cal = 8192U * (DT_INST_PROP(inst, current_lsb_microamps) *	\
+			DT_INST_PROP(inst, rshunt_micro_ohms) / 1000U) / 10000000U,\
 		.alert_config = DT_INST_PROP_OR(inst, alert_config, 0x01),	\
 		.alert_gpio = GPIO_DT_SPEC_INST_GET_OR(inst, alert_gpios, {0}),	\
 	};							    \
diff --git a/drivers/sensor/ina23x/ina237.h b/drivers/sensor/ina23x/ina237.h
index b028f5ffb9..e923bbaf05 100644
--- a/drivers/sensor/ina23x/ina237.h
+++ b/drivers/sensor/ina23x/ina237.h
@@ -50,6 +50,7 @@ struct ina237_config {
 	uint16_t adc_config;
 	uint32_t current_lsb;
 	uint32_t rshunt;
+	uint16_t cal;
 	const struct gpio_dt_spec alert_gpio;
 	uint16_t alert_config;
 };
```

I still noticed that it didn't pass clang formatting. Since this doesn't seem to be providing any extra value and I'm performing the correct order of operations any, I think I will leave the current PR the way it is.

xyzzy42 replied:
Likewise, using 64-bit math allows removing the possibility of overflow without compromising the precision of the result.

```c
#define INA237_CAL_SCALING 8192ULL
.cal = (INA237_CAL_SCALING * DT_INST_PROP(inst, current_lsb_microamps) * DT_INST_PROP(inst, rshunt_micro_ohms)) / 10000000000ULL;
```

bperseghetti replied:
Ok, yep, makes sense. Added that and listed you as co-author.

carlescufi commented:
@bperseghetti could you please first answer the remaining review comments by @xyzzy42 ?

bperseghetti commented:
> @bperseghetti could you please first answer the remaining review comments by @xyzzy42 ?

Yep, think it should all be good now.

bperseghetti committed:
drivers: sensor: ina23x: calculate cal at compile.

Calculate the calibration value at compile for ina23x.
Maximizes the precision of the calcualtion value by
using 64bit math at compile, allows for removal
of rshunt config option.

Code cleaned up with clang-format.

Co-authored-by: Trent Piepho <tpiepho@gmail.com>
Signed-off-by: Benjamin Perseghetti <bperseghetti@rudislabs.com>

xyzzy42 commented on drivers/sensor/ina23x/ina230.c:
```diff
@@ -261,32 +249,28 @@ static const struct sensor_driver_api ina230_driver_api = {
+		.bus = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.config = DT_INST_PROP(inst, config),                                              \
+		.current_lsb = DT_INST_PROP(inst, current_lsb_microamps),                          \
+		.rshunt = DT_INST_PROP(inst, rshunt_micro_ohms),                                   \
```
`rshunt` is now unused.

bperseghetti replied:
Good point, removed and pushed.

bperseghetti commented:
Tested on an INA230 and INA226 with an mimxrt1062 and mimxrt1170. Now works as desired/expected.

xyzzy42 commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -591,7 +591,7 @@ test_i2c_ina230: ina230@5e {
 	config = <0>;
 	current-lsb-microamps = <1000>;
-	rshunt-milliohms = <0>;
+	rshunt-micro-ohms = <1000>;
```
Looks like the undetected run-time divide by zero error in the test case turned into a detected compile time error.

bperseghetti replied:
yep, hence the fix of actually setting it.

xyzzy42 commented on dts/bindings/sensor/ti,ina23x-common.yaml:
```diff
@@ -34,10 +34,10 @@ properties:
       so that low maximum currents can be measured with enough resolution.
 
-  rshunt-milliohms:
+  rshunt-micro-ohms:
```
The dash looked odd here to me.  So I checked other bindings, I didn't find any of the form "prefix-unit".  It's always "microvolt", "milliamp", etc.  And a handful of "mv" and a single "uv".

bperseghetti replied:
So I followed what I was told to be the convention: https://github.com/zephyrproject-rtos/zephyr/pull/59861#discussion_r1247812946

bperseghetti replied:
To your point though I originally did have it without a dash.

xyzzy42 replied:
Indeed it did come up, sorry, I missed that comment.  And dt-schema does say micro-ohm is the standard.  Odd choice, IMHO.  The description of "micro-ohm" being "microohm" is quite perplexing.

```
  "-micro-ohms$":
    $ref: types.yaml#/definitions/uint32-array
    description: microohm
  "-microwatt$":
    $ref: types.yaml#/definitions/uint32-array
    description: microwatt
  "-milliwatt$":
    $ref: types.yaml#/definitions/uint32-array
    description: milliwatt
```

bperseghetti replied:
No worries, thanks for looking it up though!

bperseghetti commented:
@bbilas @gmarull do you guys mind re-approving now that it has the requested changes added?

bperseghetti commented:
Thanks @bbilas and @vaishnavachath! Are you able to merge this in now @carlescufi?
